src/python/pants/engine/legacy/graph.py

```diff
@@ -329,25 +341,16 @@ def create_legacy_graph_tasks(symbol_table_cls):
   """Create tasks to recursively parse the legacy graph."""
   symbol_table_constraint = symbol_table_cls.constraint()
   return [
-    # Recursively requests HydratedTargets, which will result in an eager, transitive graph walk.
-    (HydratedTargets,
-     [SelectTransitive(HydratedTarget,
-                       Addresses,
-                       field_types=(BuildFileAddress,))],
-     HydratedTargets),
-    (HydratedTarget,
-     [Select(symbol_table_constraint),
-      SelectDependencies(HydratedField,
-                         symbol_table_constraint,
-                         'field_adaptors',
-                         field_types=(SourcesField, BundlesField,))],
-     hydrate_target),
-    (HydratedField,
-     [Select(SourcesField),
-      SelectProjection(Snapshot, PathGlobs, ('path_globs',), SourcesField)],
-     hydrate_sources),
-    (HydratedField,
-     [Select(BundlesField),
-      SelectDependencies(Snapshot, BundlesField, 'path_globs_list', field_types=(PathGlobs,))],
-     hydrate_bundles),
+    transitive_hydrated_targets,
+    TaskRule(
+      HydratedTarget,
+      [Select(symbol_table_constraint),
+       SelectDependencies(HydratedField,
```

Comment:
Yea, mentioned in the description: in cases where constraints are dynamic rather than static. The symbol table isn't known until the Scheduler is constructed.